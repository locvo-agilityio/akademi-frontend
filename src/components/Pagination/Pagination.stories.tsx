import type { Meta, StoryObj } from '@storybook/react';

// Components
import Pagination from '.';

const meta: Meta<typeof Pagination> = {
  title: 'Components/Pagination',
  tags: ['autodocs'],
  component: Pagination,
};

export default meta;
type Story = StoryObj<typeof Pagination>;

export const Default: Story = {
  args: {
    totalRecords: '100 items',
    pageSize: 10,
    currentButtons: [1, 2, 3, '...', 7],
    currentPage: 2,
  },
};
