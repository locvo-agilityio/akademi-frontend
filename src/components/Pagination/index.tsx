'use client';

import { memo, useMemo } from 'react';
import { Button, Flex, Icon, Text } from '@chakra-ui/react';

// Icons
import { ArrowIcon } from '../icons';

interface PaginationProps {
  totalRecords?: string;
  pageSize?: number;
  currentPage?: number;
  isDisabledPrev?: boolean;
  isDisableNext?: boolean;
  currentButtons?: (number | string)[];
  onPageChange?: (direction: string) => void;
  onClickPage?: (currentPage: number) => void;
}

const Pagination = ({
  totalRecords,
  currentPage = 1,
  currentButtons = [],
  isDisabledPrev,
  isDisableNext,
  onPageChange,
  onClickPage,
}: PaginationProps) => {
  const handleNextPage = () => onPageChange?.('next');

  const handlePrevPage = () => onPageChange?.('prev');

  const renderBtnPage = useMemo(
    () =>
      currentButtons.map((item: string | number) => {
        const isDots = item.toString() === '...';
        const isCurrentPage = currentPage === item;
        const isDisable = isDots || isCurrentPage;
        const bgBtn = isCurrentPage ? 'primary' : 'transparent';
        const colorBtn = isCurrentPage ? 'white' : 'whiteSmoke';
        const cursor = isDots ? 'not-allowed' : '';

        const hoverStyle = !isDots
          ? {
              color: 'white',
              bg: 'primary',
            }
          : {};

        const disableStyle = !isDots
          ? {
              cursor: isCurrentPage ? 'not-allowed' : '',
              color: 'white',
              bg: 'primary',
            }
          : {};

        const handleClickPage = () => {
          if (!isDots && !isCurrentPage) {
            onClickPage?.(item as number);
          }
        };

        return (
          <Button
            key={item}
            title={`page-${item}`}
            isDisabled={isDisable}
            variant="icon"
            w={10}
            h={10}
            fontSize="base"
            textAlign="center"
            bg={bgBtn}
            color={colorBtn}
            border="2px solid"
            borderColor="whitesmoke"
            cursor={cursor}
            _hover={hoverStyle}
            _disabled={disableStyle}
            onClick={handleClickPage}
          >
            {item}
          </Button>
        );
      }),
    [currentButtons, currentPage, onClickPage],
  );

  return (
    <Flex w="full" justifyContent="space-between" mt={8}>
      <Flex alignItems="center">
        <Text fontSize="base" fontWeight="regular" color="gray.600">
          {totalRecords}
        </Text>
      </Flex>
      <Flex justifyContent="space-between" alignItems="center">
        <Button
          title="prev-button"
          w={8}
          variant="icon"
          isDisabled={isDisabledPrev}
          onClick={handlePrevPage}
          bgColor="transparent"
          border="none"
        >
          <Icon
            as={ArrowIcon}
            boxSize={4}
            transform="rotate(90deg)"
            color="whiteSmoke"
          />
        </Button>
        <Flex alignItems="center" gap={2}>
          {renderBtnPage}
        </Flex>
        <Button
          title="next-button"
          w={8}
          variant="icon"
          isDisabled={isDisableNext}
          onClick={handleNextPage}
          bgColor="transparent"
          border="none"
        >
          <Icon
            as={ArrowIcon}
            boxSize={4}
            transform="rotate(-90deg)"
            color="whiteSmoke"
          />
        </Button>
      </Flex>
    </Flex>
  );
};

export default memo(Pagination);
