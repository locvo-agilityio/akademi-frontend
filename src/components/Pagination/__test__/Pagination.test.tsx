import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

// Components
import Pagination from '..';

const onPageChangeMock = jest.fn();
const onClickPageMock = jest.fn();

describe('Pagination render', () => {
  it('Should match snapshot.', () => {
    const { container } = render(
      <Pagination
        currentPage={1}
        isDisableNext={true}
        isDisabledPrev={true}
        currentButtons={[1, 2, 3, '...', 4]}
        onClickPage={onClickPageMock}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Handle click page', async () => {
    const { getByTitle } = render(
      <Pagination onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByTitle('next-button');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('Handle next page', async () => {
    const { getByTitle } = render(
      <Pagination onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByTitle('next-button');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('Handle click number page', async () => {
    const { getByTitle } = render(
      <Pagination
        currentPage={1}
        isDisableNext={true}
        isDisabledPrev={true}
        currentButtons={[1, 2, 3, '...', 4]}
        onClickPage={onClickPageMock}
      />,
    );

    const nextPage = getByTitle('page-2');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('Handle prev page', async () => {
    const { getByTitle } = render(
      <Pagination currentPage={2} onPageChange={onPageChangeMock} />,
    );

    const nextPage = getByTitle('prev-button');

    await userEvent.click(nextPage);
    expect(onPageChangeMock).toHaveBeenCalled();
  });

  it('should render default value correctly', () => {
    const { container } = render(<Pagination />);

    expect(container).toBeInTheDocument();
  });
});
