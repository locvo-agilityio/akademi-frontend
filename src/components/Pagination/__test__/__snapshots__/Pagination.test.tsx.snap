// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination render Should match snapshot. 1`] = `
<div>
  <div
    class="css-13mwgsm"
  >
    <div
      class="css-70qvj9"
    >
      <p
        class="chakra-text css-ghb6lk"
      />
    </div>
    <div
      class="css-69i1ev"
    >
      <button
        class="chakra-button css-1umnp0x"
        disabled=""
        title="prev-button"
        type="button"
      >
        <svg
          class="chakra-icon css-1r59igl"
          fill="none"
          focusable="false"
          height="14"
          viewBox="0 0 21 14"
          width="21"
        >
          <path
            d="M20.6761 2.48803L11.2361 12.952C10.8521 13.4 10.1481 13.4 9.76415 12.952L0.324149 2.48803C-0.283851 1.81603 0.164149 0.728027 1.06015 0.728027L19.9401 0.728027C20.8361 0.728027 21.2841 1.81603 20.6761 2.48803Z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="css-j2at52"
      >
        <button
          class="chakra-button css-18o81ng"
          disabled=""
          title="page-1"
          type="button"
        >
          1
        </button>
        <button
          class="chakra-button css-c1bazq"
          title="page-2"
          type="button"
        >
          2
        </button>
        <button
          class="chakra-button css-c1bazq"
          title="page-3"
          type="button"
        >
          3
        </button>
        <button
          class="chakra-button css-1541j2i"
          disabled=""
          title="page-..."
          type="button"
        >
          ...
        </button>
        <button
          class="chakra-button css-c1bazq"
          title="page-4"
          type="button"
        >
          4
        </button>
      </div>
      <button
        class="chakra-button css-1umnp0x"
        disabled=""
        title="next-button"
        type="button"
      >
        <svg
          class="chakra-icon css-178y3kn"
          fill="none"
          focusable="false"
          height="14"
          viewBox="0 0 21 14"
          width="21"
        >
          <path
            d="M20.6761 2.48803L11.2361 12.952C10.8521 13.4 10.1481 13.4 9.76415 12.952L0.324149 2.48803C-0.283851 1.81603 0.164149 0.728027 1.06015 0.728027L19.9401 0.728027C20.8361 0.728027 21.2841 1.81603 20.6761 2.48803Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
